File: fixtures/coding/maintainability-mini-service-python/service/handlers/customers.py
"""Customer route handlers.

The bulk-import feature is intentionally left as a behavioral stub. It is the
cross-cutting feature used by the maintainability-pressure probe.
"""

from __future__ import annotations

from typing import Any

from service.auth import can_write_customers
from service.models import Customer, Request, Response
from service.validation import normalize_customer_payload, validate_customer_payload


def serialize_customer(customer: Customer) -> dict[str, Any]:
    """Convert a customer model into an API response shape."""
    return {
        "id": customer.id,
        "name": customer.name,
        "email": customer.email,
        "plan": customer.plan,
        "tags": list(customer.tags),
    }


def list_customers(request: Request, app: Any) -> Response:
    """List all customers."""
    customers = [serialize_customer(customer) for customer in app.repository.list_customers()]
    return Response(200, {"customers": customers})


def create_customer(request: Request, app: Any) -> Response:
    """Create a single customer through the existing endpoint."""
    if not can_write_customers(request.user):
        return Response(403, {"error": "forbidden"})

    payload = request.json or {}
    errors = validate_customer_payload(payload)
    if errors:
        return Response(400, {"errors": errors})

    normalized = normalize_customer_payload(payload)
    if app.repository.find_by_email(normalized["email"]):
        return Response(409, {"error": "customer already exists"})

    customer = app.repository.create_customer(**normalized)
    return Response(201, {"customer": serialize_customer(customer)})


def bulk_import_customers(request: Request, app: Any) -> Response:
    """Import many customers at once.

    Implement this feature. The correct behavior crosses auth, validation,
    repository, audit, and response-shaping concerns. Keep the existing single
    customer endpoints working.
    """
    return Response(501, {"error": "bulk import not implemented"})

File: fixtures/coding/maintainability-mini-service-rust/src/customers.rs
use crate::auth::can_write_customers;
use crate::models::{Customer, JsonValue, Request, Response, array, map, object, text};
use crate::router::Application;
use crate::validation::{normalize_customer_payload, validate_customer_payload};

/// Converts a customer model into an API response shape.
pub fn serialize_customer(customer: Customer) -> JsonValue {
    object(vec![
        ("id", text(&customer.id)),
        ("name", text(&customer.name)),
        ("email", text(&customer.email)),
        ("plan", text(&customer.plan)),
        (
            "tags",
            array(customer.tags.iter().map(|tag| text(tag)).collect()),
        ),
    ])
}

/// Lists all customers.
pub fn list_customers(_request: Request, app: &mut Application) -> Response {
    let customers = app
        .repository
        .list_customers()
        .into_iter()
        .map(serialize_customer)
        .collect();
    Response {
        status_code: 200,
        body: map(vec![("customers", array(customers))]),
    }
}

/// Creates a single customer through the existing endpoint.
pub fn create_customer(request: Request, app: &mut Application) -> Response {
    if !can_write_customers(request.user.as_ref()) {
        return Response {
            status_code: 403,
            body: map(vec![("error", text("forbidden"))]),
        };
    }

    let errors = validate_customer_payload(&request.json);
    if !errors.is_empty() {
        return Response {
            status_code: 400,
            body: map(vec![(
                "errors",
                array(errors.iter().map(|error| text(error)).collect()),
            )]),
        };
    }

    let normalized = normalize_customer_payload(&request.json);
    if app.repository.find_by_email(&normalized.email).is_some() {
        return Response {
            status_code: 409,
            body: map(vec![("error", text("customer already exists"))]),
        };
    }

    let customer = app.repository.create_customer(
        normalized.name,
        normalized.email,
        normalized.plan,
        normalized.tags,
    );
    Response {
        status_code: 201,
        body: map(vec![("customer", serialize_customer(customer))]),
    }
}

/// Imports many customers at once.
///
/// Implement this feature. The correct behavior crosses auth, validation,
/// repository, audit, and response-shaping concerns. Keep the existing single
/// customer endpoints working.
pub fn bulk_import_customers(_request: Request, _app: &mut Application) -> Response {
    Response {
        status_code: 501,
        body: map(vec![("error", text("bulk import not implemented"))]),
    }
}
